src/goto-programs/character_refine_preprocess.cpp

```diff
@@ -0,0 +1,2564 @@
+
+exprt character_refine_preprocesst::in_interval_expr(
+  const exprt &chr, mp_integer lower_bound, mp_integer upper_bound)
+{
+  and_exprt res(
+    binary_relation_exprt(chr, ID_ge, from_integer(lower_bound, chr.type())),
+    binary_relation_exprt(chr, ID_le, from_integer(upper_bound, chr.type())));
+  return res;
+}
+
+/*******************************************************************\
+
+Function: character_refine_preprocesst::in_list_expr
+
+  Inputs:
+    chr - An expression of type character
+    list - A list of integer representing unicode characters
+
+ Outputs: A Boolean expression
+
+ Purpose: The returned expression is true when the given character
+          is equal to one of the element in the list
+
+\*******************************************************************/
+
+exprt character_refine_preprocesst::in_list_expr(
+  const exprt &chr, const std::list<mp_integer> &list)
+{
+  exprt res=false_exprt();
+  for(auto i : list)
+    res=or_exprt(res, equal_exprt(chr, from_integer(i, chr.type())));
+  return res;
+}
+
+/*******************************************************************\
+
+Function: character_refine_preprocesst::expr_of_char_count
+
+  Inputs:
+    expr - An expression of type character
+    type - A type for the output
+
+ Outputs: A integer expression of the given type
+
+ Purpose: Determines the number of char values needed to represent
+          the specified character (Unicode code point).
+
+\*******************************************************************/
+
+exprt character_refine_preprocesst::expr_of_char_count(
+  const exprt &chr, const typet &type)
+{
+  exprt u010000=from_integer(0x010000, type);
+  binary_relation_exprt small(chr, ID_lt, u010000);
+  return if_exprt(small, from_integer(1, type), from_integer(2, type));
+}
+
+/*******************************************************************\
+
+Function: character_refine_preprocesst::convert_char_count
+
+  Inputs:
+    target - a position in a goto program
+
+ Purpose: Converts function call to an assignment of an expression
+          corresponding to the java method Character.charCount:(I)I
+
+\*******************************************************************/
+
+void character_refine_preprocesst::convert_char_count(conversion_input &target)
+{
+  convert_char_function(
+    &character_refine_preprocesst::expr_of_char_count, target);
+}
+
+/*******************************************************************\
+
+Function: character_refine_preprocesst::expr_of_char_value
+  Inputs:
+    expr - An expression of type character
+    type - A type for the output
+
+ Outputs: An expression of the given type
+
+ Purpose: Casts the given expression to the given type
+
+\*******************************************************************/
+
+exprt character_refine_preprocesst::expr_of_char_value(
+  const exprt &chr, const typet &type)
+{
+  return typecast_exprt(chr, type);
+}
```

Comment:
Actually I need a reference to a function with the signature `exprt (*expr_function)(const exprt &chr, const typet &type)` in order to pass it to `convert_char_function`. That's why I have to declare this function.